cgen: make `.global.`s work more according to spec

zerbina opened the pull request:
## Summary
For the C target, globals defined inside procedures that are resource- like (i.e. have a user-defined or lifted destructor) are now initialized in a module's pre-init procedure, instead of each time control-flow reaches the definition. This mirrors what already happened for non-resource-like globals.

In addition, globals defined in `inline` procedures no longer cause linking errors because of duplicated definitions in the generated C code.

## Details
The globals are extracted before translating the AST to MIR code, making the workaround in `mirgen` (that didn't work anyway) obsolete. In order to work towards unifying the architecture of the back-ends, `vmbackend` now also makes use of the pre-extraction, no longer requiring extra logic in the code-generator.

### Known Issues
- the destructor-injection pass is not run for the initializer expression of globals defined inside procedures
- `.global.`s on the JS target still don't work according to specification

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

## Notes for Reviewers
The PR is meant to unblock #450 with the least amount of work and changes to semantics. I think that the introduced regression (no destructor injection for initializers of procedure-level globals) is acceptable until the implementation gets an overhaul.

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

saem commented on compiler/sem/injectdestructors.nim:
```diff
@@ -1103,6 +1103,14 @@ func shouldInjectDestructorCalls*(owner: PSym): bool =
 
+proc deferGlobalDestructor*(g: ModuleGraph, idgen: IdGenerator, owner: PSym,
+                            global: PNode) =
+  ## If the global has a destructor, emits a call to it to the end of the
```
```suggestion
  ## If the global has a destructor, emits a call to it at the end of the
```

saem commented on compiler/sem/transf.nim:
```diff
@@ -1154,3 +1154,60 @@ proc transformExpr*(g: ModuleGraph; idgen: IdGenerator; module: PSym, n: PNode):
+  of nkWhen:
+    # a ``when nimvm`` construct
+    # XXX: this logic duplciates what ``mirgen`` already does. Maybe
+    #      collecting should happen there? Or procedure-level globals should
```
Likely makes sense in sem as a normalization step.

saem commented on compiler/sem/transf.nim:
```diff
@@ -1154,3 +1154,60 @@ proc transformExpr*(g: ModuleGraph; idgen: IdGenerator; module: PSym, n: PNode):
+    extractGlobals(body[1], output, isNimVm)
+  of nkWhen:
+    # a ``when nimvm`` construct
+    # XXX: this logic duplciates what ``mirgen`` already does. Maybe
```
```suggestion
    # XXX: this logic duplicates what ``mirgen`` already does. Maybe
```

saem commented on compiler/sem/transf.nim:
```diff
@@ -1154,3 +1154,60 @@ proc transformExpr*(g: ModuleGraph; idgen: IdGenerator; module: PSym, n: PNode):
+
+    extractGlobals(branch, output, isNimVm)
+  of nkVarSection, nkLetSection:
+    # iterated over all children and extract identdefs of globals:
```
```suggestion
    # iterates over all children and extract identdefs of globals:
```

saem commented on compiler/vm/vmbackend.nim:
```diff
@@ -177,8 +177,13 @@ proc generateTopLevelStmts*(module: var Module, c: var TCtx,
+proc generateCodeForProc(c: var TCtx, s: PSym,
+                         globals: var seq[PNode]): VmGenResult =
+  ## Generates and emits the bytecode for the procedure `s`. The globals
+  ## defined in it it are extracted from the body and their identdefs appended
```
```suggestion
  ## defined in it are extracted from the body and their identdefs appended
```

zerbina committed:
cgen: make `.global.`s work more according to spec

For the C target, globals defined inside procedures that are resource-
like (i.e. have a user-defined or lifted destructor) are now
initialized in a module's pre-init procedure, instead of each time
control-flow reaches the definition. This mirrors what already happened
for non-resource-like globals.

In addition, globals defined in `inline` procedures no longer cause
linking errors because of duplicated definitions in the generated C
code.

The globals are extracted before translating the AST to MIR code,
making the workaround in `mirgen` (that didn't work anyway) obsolete.
In order to work towards unifying the architecture of the back-ends,
`vmbackend` now also makes use of the pre-extraction, no longer
requiring extra logic in the code-generator.

- the destructor injection pass is not run for the initializer
  expression of globals defined inside procedures
- `.global.`s on the JS target still don't work according to
  specification

saem reviewed (approved):
Bors r+

Thanks for the detailed explanation, it was really easy to follow.

saem commented:
bors r+
